[analyzer][NFC] Turn NodeBuilderContext into a class

sona-tau opened the pull request:
From issue #73088. I changed `NodeBuilderContext` into a class. Additionally, there were some other mentions of the former being a struct which I also changed into a class. This is my first time working with an issue so I will be open to hearing any advice or changes that need to be done.

sona-tau committed:
[analyzer] Turn NodeBuilderContext into a class

steakhal commented:
Thanks for the PR.
I think it would make sense to not expose data members in a class' public api.
We might need some member functions along the way to make this possible.

steakhal reviewed (changes requested):
See my previous reply.

sona-tau committed:
fix(NodeBuilderContext implementation): changed class members to private and added the appropriate getters.

Moved every member in NodeBuilderContext to private and added the following getters:
- getEngine
- getLocationContext

Setters are not necessary since the class members are const already.

diego-est commented:
I added the appropriate functions and moved the class members into private. Additionally I ran `make check-all` to make sure the now-private members wouldn't cause any issues with the rest of the codebase, all the checks passed.

steakhal commented on clang/include/clang/StaticAnalyzer/Core/PathSensitive/CoreEngine.h:
```diff
@@ -194,11 +194,12 @@ class CoreEngine {
 };
 
 // TODO: Turn into a class.
```
Remove this TODO, as its completed.

diego-est replied:
Done.

steakhal reviewed (commented):
Alright. So we can simply mark all the members private. I was surprised a bit.
LGTM, but remove the TODO comment you fix.

sona-tau committed:
fix(NodeBuilderContext API): fixed accesses to NodeBuilderContext private member accesses.

The files CheckerManager.h, CoreEngine.h, ExprEngine.h and CoreEngine.cpp all had bad private member accesses.

sona-tau committed:
fix(CoreEngine.cpp): fixed type in ternary

Fixed code formatting error in ternary expression.

diego-est commented:
Seems like it wasn't as easy as just making them private. I think I found all the places where there were private member accesses. The tests passed on my side again and the github-pull-requests action. The code_formatter action keeps failing because of the documentation on something I didn't touch (but seems to be fixed high upstream). Everything should be fixed now.
